### PR TITLE
Tags

### PR DESCRIPTION
**Pull request by lukedirtwalker**
This is still Work in progress but I would like to get Feedback about the general structure
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23discussion_r40246609%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23discussion_r40246730%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23discussion_r40247088%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23discussion_r40247200%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23discussion_r40247372%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23discussion_r40247483%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23discussion_r40247713%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23discussion_r40247799%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23issuecomment-142708292%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23issuecomment-142883647%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23issuecomment-142889508%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23issuecomment-142889872%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23issuecomment-142890369%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23issuecomment-142891460%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23issuecomment-142708292%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20like%20it.%20There%20is%20more%20to%20be%20done%20in%20terms%20of%20expanding%20the%20queries%2C%20but%20it%27s%20definitely%20going%20in%20the%20right%20direction.%20If%20possible%2C%20could%20you%20please%20fix%20all%20the%20tiny%20issues%20that%20I%20mention%20here%20and%20I%27ll%20accept%20the%20PR%2C%20even%20in%20this%20incomplete%20state.%20Then%20at%20least%20I%27d%20only%20have%20to%20look%20at%20the%20new%20changes%20that%20come%20afterwards.%5Cr%5Cn%5Cr%5CnWe%20should%20meet%20to%20discuss%20your%20idea%20about%20tags.%20Keep%20in%20mind%20that%20tags%2C%20in%20my%20mind%20will%20be%20primarily%20used%20with%20your%20queries%2C%20so%20we%20need%20to%20have%20convenient%20ways%20to%20query%20and%20edit%20them.%5Cr%5Cn%5Cr%5CnI%20might%20have%20some%20time%20to%20meet%20tomorrow%20%28Thursday%29%2C%20but%20I%20can%27t%20say%20for%20sure%20at%20the%20moment.%22%2C%20%22created_at%22%3A%20%222015-09-23T19%3A45%3A17Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20merged%20this%20but%20I%27m%20wondering%2C%20why%20do%20we%20need%20a%20tag%20node%3F%20If%20we%20decide%20in%20the%20end%20that%20tags%20are%20just%20a%20list%20of%20strings%2C%20we%20should%20directly%20register%20a%20List%20of%20Strings%20attribute%20as%20part%20of%20the%20extension%2C%20and%20not%20bother%20with%20having%20a%20separate%20node.%20This%20way%2C%20you%20can%20even%20make%20that%20a%20required%20attribute%2C%20and%20missing%20lists%20will%20be%20autmatically%20empty.%22%2C%20%22created_at%22%3A%20%222015-09-24T10%3A23%3A24Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Hm%20well%20previously%20the%20idea%20was%20to%20have%20a%20bit%20more%20structured%20data.%20But%20now%20you%20are%20right.%5Cr%5Cn%5Cr%5CnOn%20the%20other%20hand%20one%20thing%20we%20should%20definitely%20support%20is%20having%20multiple%20tags%20at%20one%20node.%20Maybe%20the%20extension%20is%20then%20just%20a%20list%20of%20TagNode%20or%20do%20you%20have%20a%20better%20approach%20to%20solve%20this%3F%22%2C%20%22created_at%22%3A%20%222015-09-24T10%3A46%3A45Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%2C%20%7B%22body%22%3A%20%22But%20isn%27t%20a%20List%20of%20String%20%28as%20opposed%20to%20a%20List%20of%20TagNode%29%20enough%3F%20If%20we%20decide%20that%20we%20don%27t%20need%20more%20structure%20than%20a%20string%2C%20then%20we%20can%20completely%20get%20rid%20of%20the%20TagNode%20type.%22%2C%20%22created_at%22%3A%20%222015-09-24T10%3A48%3A49Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Actually%20I%20think%20a%20set%20would%20make%20more%20sense%3F%22%2C%20%22created_at%22%3A%20%222015-09-24T10%3A51%3A10Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%2C%20%7B%22body%22%3A%20%22Well%20yes%2C%20but%20we%20don%27t%20have%20set%20attributes.%20We%20can%20use%20a%20list%20and%20just%20ignore%20the%20order.%22%2C%20%22created_at%22%3A%20%222015-09-24T10%3A53%3A14Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20a08c5e3947f975973d6d221fa0fea9e1b27dbf11%20InformationScripting/src/queries/TagQuery.cpp%2036%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23discussion_r40247372%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22You%20already%20talk%20about%20types%20and%20names%20of%20tags%2C%20but%20I%20don%27t%20know%20what%20these%20are.%20We%20should%20talk%20about%20your%20design.%22%2C%20%22created_at%22%3A%20%222015-09-23T19%3A36%3A59Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/TagQuery.cpp%3AL1-125%22%7D%2C%20%22Pull%20a08c5e3947f975973d6d221fa0fea9e1b27dbf11%20InformationScripting/src/queries/ScopedArgumentQuery.h%2056%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23discussion_r40247088%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Use%20default%20initializer%20here%20%60%7B%7D%60.%22%2C%20%22created_at%22%3A%20%222015-09-23T19%3A34%3A17Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/ScopedArgumentQuery.h%3AL1-65%22%7D%2C%20%22Pull%20a08c5e3947f975973d6d221fa0fea9e1b27dbf11%20InformationScripting/src/queries/TagQuery.cpp%20116%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23discussion_r40247799%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22It%27s%20more%20performant%20to%20call%20begin/end%20modification%20only%20once%20for%20the%20entire%20method%2C%20and%20for%20each%20individual%20node%20to%20just%20call%20%60manger%28%29-%3EchangeModificationTarget%28%29%60.%22%2C%20%22created_at%22%3A%20%222015-09-23T19%3A40%3A56Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/TagQuery.cpp%3AL1-125%22%7D%2C%20%22Pull%20a08c5e3947f975973d6d221fa0fea9e1b27dbf11%20InformationScripting/src/queries/TagQuery.cpp%2066%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23discussion_r40247483%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Why%20are%20%60Local%60%20and%20%60Global%60%20empty%3F%20Doesn%27t%20it%20make%20sense%20to%20just%20get%20all%20the%20tags%2C%20of%20let%27s%20say%20the%20current%20selection%3F%22%2C%20%22created_at%22%3A%20%222015-09-23T19%3A38%3A04Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/TagQuery.cpp%3AL1-125%22%7D%2C%20%22Pull%20a08c5e3947f975973d6d221fa0fea9e1b27dbf11%20InformationScripting/src/nodes/TagNode.h%2040%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23discussion_r40246730%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22This%20should%20be%20an%20attribute.%20Why%20isn%27t%20it%3F%22%2C%20%22created_at%22%3A%20%222015-09-23T19%3A31%3A01Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/nodes/TagNode.h%3AL1-51%22%7D%2C%20%22Pull%20a08c5e3947f975973d6d221fa0fea9e1b27dbf11%20InformationScripting/src/queries/TagQuery.cpp%20114%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23discussion_r40247713%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Brilliant.%20You%27ve%20just%20implemented%20your%20first%20query%20that%20does%20modifications.%20I%20like%20it.%20I%20like%20how%20natural%20it%20feels%20to%20add%20tags%20this%20way.%20There%20might%20be%20other%20ways%2C%20but%20this%20one%20feels%20very%20natural%2C%20especially%20for%20adding%20tags%20on%20mass.%22%2C%20%22created_at%22%3A%20%222015-09-23T19%3A39%3A58Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/TagQuery.cpp%3AL1-125%22%7D%2C%20%22Pull%20a08c5e3947f975973d6d221fa0fea9e1b27dbf11%20InformationScripting/src/nodes/TagExtension.h%2032%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23discussion_r40246609%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Do%20you%20need%20to%20include%20this%3F%20You%20don%27t%20seem%20to%20use%20%60Text%60%20anywhere.%22%2C%20%22created_at%22%3A%20%222015-09-23T19%3A29%3A44Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/nodes/TagExtension.h%3AL1-48%22%7D%2C%20%22Pull%20a08c5e3947f975973d6d221fa0fea9e1b27dbf11%20ModelBase/src/nodes/composite/CompositeNode.h%206%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/155%23discussion_r40247200%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22What%27s%20there%20to%20fix%3F%20We%20just%20leave%20it%20public.%20I%20think%20that%27s%20fine.%22%2C%20%22created_at%22%3A%20%222015-09-23T19%3A35%3A21Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20ModelBase/src/nodes/composite/CompositeNode.h%3AL136-153%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [x] <a href='#crh-comment-Pull a08c5e3947f975973d6d221fa0fea9e1b27dbf11 InformationScripting/src/nodes/TagExtension.h 32'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/155#discussion_r40246609'>File: InformationScripting/src/nodes/TagExtension.h:L1-48</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Do you need to include this? You don't seem to use `Text` anywhere.
- [x] <a href='#crh-comment-Pull a08c5e3947f975973d6d221fa0fea9e1b27dbf11 InformationScripting/src/nodes/TagNode.h 40'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/155#discussion_r40246730'>File: InformationScripting/src/nodes/TagNode.h:L1-51</a></b>
- <a href='https://github.com/dimitar-asenov'></a> This should be an attribute. Why isn't it?
- [x] <a href='#crh-comment-Pull a08c5e3947f975973d6d221fa0fea9e1b27dbf11 InformationScripting/src/queries/ScopedArgumentQuery.h 56'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/155#discussion_r40247088'>File: InformationScripting/src/queries/ScopedArgumentQuery.h:L1-65</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Use default initializer here `{}`.
- [x] <a href='#crh-comment-Pull a08c5e3947f975973d6d221fa0fea9e1b27dbf11 ModelBase/src/nodes/composite/CompositeNode.h 6'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/155#discussion_r40247200'>File: ModelBase/src/nodes/composite/CompositeNode.h:L136-153</a></b>
- <a href='https://github.com/dimitar-asenov'></a> What's there to fix? We just leave it public. I think that's fine.
- [ ] <a href='#crh-comment-Pull a08c5e3947f975973d6d221fa0fea9e1b27dbf11 InformationScripting/src/queries/TagQuery.cpp 36'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/155#discussion_r40247372'>File: InformationScripting/src/queries/TagQuery.cpp:L1-125</a></b>
- <a href='https://github.com/dimitar-asenov'></a> You already talk about types and names of tags, but I don't know what these are. We should talk about your design.
- [x] <a href='#crh-comment-Pull a08c5e3947f975973d6d221fa0fea9e1b27dbf11 InformationScripting/src/queries/TagQuery.cpp 66'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/155#discussion_r40247483'>File: InformationScripting/src/queries/TagQuery.cpp:L1-125</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Why are `Local` and `Global` empty? Doesn't it make sense to just get all the tags, of let's say the current selection?
- [x] <a href='#crh-comment-Pull a08c5e3947f975973d6d221fa0fea9e1b27dbf11 InformationScripting/src/queries/TagQuery.cpp 114'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/155#discussion_r40247713'>File: InformationScripting/src/queries/TagQuery.cpp:L1-125</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Brilliant. You've just implemented your first query that does modifications. I like it. I like how natural it feels to add tags this way. There might be other ways, but this one feels very natural, especially for adding tags on mass.
- [x] <a href='#crh-comment-Pull a08c5e3947f975973d6d221fa0fea9e1b27dbf11 InformationScripting/src/queries/TagQuery.cpp 116'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/155#discussion_r40247799'>File: InformationScripting/src/queries/TagQuery.cpp:L1-125</a></b>
- <a href='https://github.com/dimitar-asenov'></a> It's more performant to call begin/end modification only once for the entire method, and for each individual node to just call `manger()->changeModificationTarget()`.
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/155#issuecomment-142708292'>General Comment</a></b>
- <a href='https://github.com/dimitar-asenov'></a> I like it. There is more to be done in terms of expanding the queries, but it's definitely going in the right direction. If possible, could you please fix all the tiny issues that I mention here and I'll accept the PR, even in this incomplete state. Then at least I'd only have to look at the new changes that come afterwards.
  We should meet to discuss your idea about tags. Keep in mind that tags, in my mind will be primarily used with your queries, so we need to have convenient ways to query and edit them.
  I might have some time to meet tomorrow (Thursday), but I can't say for sure at the moment.
- <a href='https://github.com/dimitar-asenov'></a> I merged this but I'm wondering, why do we need a tag node? If we decide in the end that tags are just a list of strings, we should directly register a List of Strings attribute as part of the extension, and not bother with having a separate node. This way, you can even make that a required attribute, and missing lists will be autmatically empty.
- <a href='https://github.com/lukedirtwalker'></a> Hm well previously the idea was to have a bit more structured data. But now you are right.
  On the other hand one thing we should definitely support is having multiple tags at one node. Maybe the extension is then just a list of TagNode or do you have a better approach to solve this?
- <a href='https://github.com/dimitar-asenov'></a> But isn't a List of String (as opposed to a List of TagNode) enough? If we decide that we don't need more structure than a string, then we can completely get rid of the TagNode type.
- <a href='https://github.com/lukedirtwalker'></a> Actually I think a set would make more sense?
- <a href='https://github.com/dimitar-asenov'></a> Well yes, but we don't have set attributes. We can use a list and just ignore the order.

<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/155?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/155?approve=1'></a>&nbsp;<a href='https://github.com/dimitar-asenov/Envision/pull/155'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by dimitar-asenov** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/dimitar-asenov/Envision/pull/155'>Refresh</a>.<a href='#crh-update-none'></a>

File: InformationScripting/src/nodes/TagExtension.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+**
+***********************************************************************************************************************/
+
+#pragma once
+
+#include "../informationscripting_api.h"
+
+#include "ModelBase/src/nodes/composite/CompositeNode.h"
+#include "ModelBase/src/nodes/Text.h"
```
Comment (dimitar-asenov) [inline InformationScripting/src/nodes/TagExtension.h]: Do you need to include this? You don't seem to use `Text` anywhere.

File: InformationScripting/src/nodes/TagNode.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+#include "ModelBase/src/nodes/composite/CompositeNode.h"
+
+namespace InformationScripting {
+
+class INFORMATIONSCRIPTING_API TagNode : public Super<Model::CompositeNode>
+{
+	COMPOSITENODE_DECLARE_STANDARD_METHODS(TagNode)
+
+	public:
+		TagNode(const QString& name);
```
Comment (dimitar-asenov) [inline InformationScripting/src/nodes/TagNode.h:42]: This should be an attribute. Why isn't it? I meant the name, below.

File: InformationScripting/src/queries/ScopedArgumentQuery.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+		ScopedArgumentQuery(Model::Node* target, std::initializer_list<QCommandLineOption> options,
+								  const QStringList& args);
+		Model::Node* target() const;
+		Scope scope() const;
+
+		QString argument(const QString& argName) const;
+
+	private:
+		std::unique_ptr<QCommandLineParser> argParser_{};
+		Model::Node* target_;
```
Comment (dimitar-asenov) [inline InformationScripting/src/queries/ScopedArgumentQuery.h]: Use default initializer here `{}`.

File: ModelBase/src/nodes/composite/CompositeNode.h [hunk 4]
```diff
@@ -136,14 +136,17 @@ class MODELBASE_API CompositeNode: public Super<Node>
 
 		const AttributeChain& meta();
 
+		template <class Extension> static void registerNewExtension();
+
+		// FIXME this was protected but for registerNewExtension to work we need it public
```
Comment (dimitar-asenov) [inline ModelBase/src/nodes/composite/CompositeNode.h]: What's there to fix? We just leave it public. I think that's fine.

File: InformationScripting/src/queries/TagQuery.cpp [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+#include "TagQuery.h"
+
+#include "QueryRegistry.h"
+#include "../nodes/TagExtension.h"
+
+#include "ModelBase/src/nodes/composite/CompositeNode.h"
+
+namespace InformationScripting {
+
+const QStringList TagQuery::TAGTYPE_ARGUMENT_NAMES{"t", "type"};
```
Comment (dimitar-asenov) [inline InformationScripting/src/queries/TagQuery.cpp:36]: You already talk about types and names of tags, but I don't know what these are. We should talk about your design.

File: InformationScripting/src/queries/TagQuery.cpp [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+			{NAME_ARGUMENT_NAMES, "Name of a symbol", NAME_ARGUMENT_NAMES[1]},
+		}, QStringList("TagQuery") + args}, exec_{exec}
+{}
+
+QList<TupleSet> TagQuery::queryTags(QList<TupleSet> input)
+{
+	QList<TupleSet> result;
+	if (scope() == Scope::Local)
+	{
+
```
Comment (dimitar-asenov) [inline InformationScripting/src/queries/TagQuery.cpp]: Why are `Local` and `Global` empty? Doesn't it make sense to just get all the tags, of let's say the current selection?

File: InformationScripting/src/queries/TagQuery.cpp [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+		Q_ASSERT(input.size() > 0);
+		TupleSet tupleSet = input.takeFirst();
+		auto astTuples = tupleSet.tuples("ast");
+		for (auto tuple : astTuples)
+		{
+			Model::Node* node = tuple["ast"];
+			if (auto astNode = DCast<Model::CompositeNode>(node))
+			{
+				auto tagExtension = astNode->extension<TagExtension>();
+				astNode->beginModification("addTag");
```
Comment (dimitar-asenov) [inline InformationScripting/src/queries/TagQuery.cpp]: Brilliant. You've just implemented your first query that does modifications. I like it. I like how natural it feels to add tags this way. There might be other ways, but this one feels very natural, especially for adding tags on mass.

File: InformationScripting/src/queries/TagQuery.cpp [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+				tagExtension->setTag(new TagNode{"foo"});
+				astNode->endModification();
```
Comment (dimitar-asenov) [inline InformationScripting/src/queries/TagQuery.cpp]: It's more performant to call begin/end modification only once for the entire method, and for each individual node to just call `manger()->changeModificationTarget()`.

**Comment by dimitar-asenov** [comment]
I like it. There is more to be done in terms of expanding the queries, but it's definitely going in the right direction. If possible, could you please fix all the tiny issues that I mention here and I'll accept the PR, even in this incomplete state. Then at least I'd only have to look at the new changes that come afterwards.

We should meet to discuss your idea about tags. Keep in mind that tags, in my mind will be primarily used with your queries, so we need to have convenient ways to query and edit them.

I might have some time to meet tomorrow (Thursday), but I can't say for sure at the moment.

**Comment by dimitar-asenov** [comment]
I merged this but I'm wondering, why do we need a tag node? If we decide in the end that tags are just a list of strings, we should directly register a List of Strings attribute as part of the extension, and not bother with having a separate node. This way, you can even make that a required attribute, and missing lists will be autmatically empty.

**Comment by lukedirtwalker** [comment]
Hm well previously the idea was to have a bit more structured data. But now you are right.

On the other hand one thing we should definitely support is having multiple tags at one node. Maybe the extension is then just a list of TagNode or do you have a better approach to solve this?

**Comment by dimitar-asenov** [comment]
But isn't a List of String (as opposed to a List of TagNode) enough? If we decide that we don't need more structure than a string, then we can completely get rid of the TagNode type.

**Comment by lukedirtwalker** [comment]
Actually I think a set would make more sense?

**Comment by dimitar-asenov** [comment]
Well yes, but we don't have set attributes. We can use a list and just ignore the order.

**Pull request merged**